File: apps/ios/BrickBack/Features/Party/PartyAvatar.swift
import SwiftUI
import BrickBackKit

/// A stable seeded colour per member (wireframe palette — recoloured in S7). Ports
/// `party_avatar.dart`. Uses a deterministic string hash so a member keeps the same colour across
/// launches (Swift's `hashValue` is per-run randomized, so it can't seed this).
private let partyPalette: [Color] = [
    AppColors.info,
    AppColors.success,
    AppColors.warning,
    AppColors.danger,
    Color(hex: 0x6D28D9),
    Color(hex: 0x0E7490),
]

/// FNV-1a over the UTF-8 bytes — small, stable, good enough to spread seeds across the palette.
private func stableHash(_ s: String) -> Int {
    var h: UInt64 = 1469598103934665603
    for byte in s.utf8 {
        h ^= UInt64(byte)
        h = h &* 1099511628211
    }
    return Int(h % UInt64(Int.max))
}

struct PartyAvatar: View {
    let member: PartyMember
    var size: CGFloat = 36

    private var seed: String { member.displayName ?? member.userId }
    private var initial: String {
        let trimmed = member.displayName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? "?" : String(trimmed.prefix(1)).uppercased()
    }
    private var color: Color { partyPalette[stableHash(seed) % partyPalette.count] }

    var body: some View {
        Text(initial)
            .font(.system(size: size * 0.42, weight: .bold))
            .foregroundStyle(AppColors.onPrimary)
            .frame(width: size, height: size)
            .background(color)
            .clipShape(Circle())
            // A host is ringed in ink; members in the card colour so they read flat.
            .overlay(Circle().stroke(member.isHost ? AppColors.ink : AppColors.card, lineWidth: 2))
    }
}

/// Overlapping avatar row with a `+N` overflow chip. Port of `AvatarStack`.
struct AvatarStack: View {
    let members: [PartyMember]
    var max: Int = 5
    var size: CGFloat = 36

    var body: some View {
        let shown = Array(members.prefix(max))
        let overflow = members.count - shown.count
        let overlap = size * 0.35
        let step = size - overlap
        let width = shown.isEmpty
            ? 0
            : size + CGFloat(shown.count - 1) * step + (overflow > 0 ? step : 0)

        ZStack(alignment: .leading) {
            ForEach(Array(shown.enumerated()), id: \.element.id) { index, member in
                PartyAvatar(member: member, size: size)
                    .offset(x: CGFloat(index) * step)
            }
            if overflow > 0 {
                Text("+\(overflow)")
                    .font(.system(size: size * 0.34, weight: .bold))
                    .foregroundStyle(AppColors.inkSoft)
                    .frame(width: size, height: size)
                    .background(AppColors.faint)
                    .clipShape(Circle())
                    .overlay(Circle().stroke(AppColors.card, lineWidth: 2))
                    .offset(x: CGFloat(shown.count) * step)
            }
        }
        .frame(width: width, height: size, alignment: .leading)
    }
}

File: apps/mobile/lib/features/party/party_avatar.dart
import 'package:flutter/widgets.dart';

import '../../theme/app_theme.dart';
import 'party_models.dart';

/// A stable seeded colour per member (wireframe palette — recoloured in Phase 9).
const _palette = [
  AppColors.info,
  AppColors.success,
  AppColors.warning,
  AppColors.danger,
  Color(0xFF6D28D9),
  Color(0xFF0E7490),
];

class PartyAvatar extends StatelessWidget {
  const PartyAvatar({super.key, required this.member, this.size = 36});
  final PartyMember member;
  final double size;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    final seed = member.displayName ?? member.userId;
    final initial = (member.displayName?.trim().isNotEmpty ?? false)
        ? member.displayName!.trim()[0].toUpperCase()
        : '?';
    final color = _palette[seed.hashCode.abs() % _palette.length];
    return Container(
      width: size,
      height: size,
      alignment: Alignment.center,
      decoration: BoxDecoration(
        color: color,
        shape: BoxShape.circle,
        // A host is ringed in ink; members in the canvas colour so they read flat.
        border: Border.all(
          color: member.isHost ? c.ink : c.card,
          width: 2,
        ),
      ),
      child: Text(
        initial,
        style: TextStyle(
          color: c.onPrimary,
          fontWeight: FontWeight.w700,
          fontSize: size * 0.42,
          height: 1.0,
        ),
      ),
    );
  }
}

/// Overlapping avatar row with a `+N` overflow chip.
class AvatarStack extends StatelessWidget {
  const AvatarStack({super.key, required this.members, this.max = 5, this.size = 36});
  final List<PartyMember> members;
  final int max;
  final double size;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    final shown = members.take(max).toList();
    final overflow = members.length - shown.length;
    final overlap = size * 0.35;
    return SizedBox(
      height: size,
      width: shown.isEmpty
          ? 0
          : size + (shown.length - 1) * (size - overlap) + (overflow > 0 ? size - overlap : 0),
      child: Stack(
        children: [
          for (var i = 0; i < shown.length; i++)
            Positioned(left: i * (size - overlap), child: PartyAvatar(member: shown[i], size: size)),
          if (overflow > 0)
            Positioned(
              left: shown.length * (size - overlap),
              child: Container(
                width: size,
                height: size,
                alignment: Alignment.center,
                decoration: BoxDecoration(
                  color: c.faint,
                  shape: BoxShape.circle,
                  border: Border.all(color: c.card, width: 2),
                ),
                child: Text('+$overflow',
                    style: TextStyle(
                        color: c.inkSoft,
                        fontWeight: FontWeight.w700,
                        fontSize: size * 0.34)),
              ),
            ),
        ],
      ),
    );
  }
}
